### ISSUE TITLE
[v3] Implement new Concept Exercise: exceptions

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/fsharp#862]
This issue describes how to implement the `exceptions` concept exercise for the F# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the basics of the Concept of [Exceptions in F#][exception-handling].

## Learning objectives

- Know what exceptions are.
- Know when an exception should be raised.
- Know how to raise (and reraise) an exception.
- Know how to catch an exception.
- Know how to use a finally block to always execute logic even when there are exceptions.
- Know the most important built-in exceptions (`ArgumentException`, `InvalidOperationException`) and built-in functions to raise them.
- Know how to define a custom exception.
- Know how to use pattern matching on types.

## Out of scope

- Memory and performance characteristics.
- Know when to use exceptions vs results.

## Concepts

The Concepts this exercise unlocks are:

- `exceptions`: know what exceptions are; know when an exception should be raised; know how to raise (and reraise) an exception; know how to catch an exception; know how to use a finally block to always execute logic even when there are exceptions; know the most important built-in exceptions (`ArgumentException`, `InvalidOperationException`) and built-in functions to raise them; know how to define a custom exception; know how to use pattern matching with exceptions; know when to use exceptions vs results.
- `pattern-matching-types`: know how to pattern match on a specific type.

## Prerequisites

This exercise's prerequisites Concepts are:

- `pattern-matching`: know how to do pattern matching.
- `string-formatting`: know how to do string formatting (used in the `failwithf` function).
- `inheritance`: know how to inherit from a class.

### Hints

- [Exception handling][exception-handling]: exception handling in F#.
- [Exception introduction][exeptions]: introduction to exceptions in F#.

### After

- [Exception handling][exception-handling]: exception handling in F#.
- [Exception introduction][exeptions]: introduction to exceptions in F#.
- [Exceptions vs Result][exceptions-vs-result]: shows the difference between using exceptions vs Results.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/fsharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/fsharp/reference/README.md#reference-docs
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[meta-design]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[exeptions]: https://fsharpforfunandprofit.com/posts/exceptions/
[exception-handling]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/exception-handling/
[exceptions-vs-result]: https://danielwertheim.se/exceptions-vs-the-result-type-in-fsharp/
